packages/components/src/components/Calculator/index.tsx

```diff
@@ -1,158 +1,237 @@
-  updateAndProcessFieldCode,
-} from "./asyncActions.js";
+  FormShape,
+  InputValues,
+  SqCalculatorValueWithContext,
+  SqValueResult,
+} from "./types.js";
+import { useSavedCalculatorState } from "./useSavedCalculatorState.js";
+
+async function runSquiggleCode(
+  code: string,
+  environment: Env
+): Promise<SqValueResult | undefined> {
+  if (!code) {
+    return undefined;
+  }
+  const project = SqProject.create();
+  if (environment) {
+    project.setEnvironment(environment);
+  }
+  const sourceId = "calculator";
+  project.setSource(sourceId, code);
+  await project.run(sourceId);
+  return project.getResult(sourceId);
+}
+
+function fieldValueToCode(
+  name: string,
+  calculator: SqCalculator,
+  fieldValue: string | boolean
+): string {
+  const input = calculator.inputs.find((row) => row.name === name);
+  if (!input) {
+    throw new Error("Invalid input name.");
+  }
+  if (typeof fieldValue === "boolean") {
+    return fieldValue.toString();
+  }
+
+  if (input.tag === "select") {
+    return `"${fieldValue}"`;
+  } else {
+    return fieldValue;
+  }
+}
+
+function getFormValues(calculator: SqCalculator): FormShape {
+  return Object.fromEntries(
+    calculator.inputs.map((input) => [input.name, input.default ?? ""])
+  );
+}
+
+/**
+ * This function implements all state flow logic for calculator inputs.
+ */
+function useCalculator(
+  valueWithContext: SqCalculatorValueWithContext,
+  environment: Env
+) {
+  // useMemo here is important; valueWithContext.value is a getter that creates a new calculator value every time.
+  const calculator = useMemo(() => valueWithContext.value, [valueWithContext]);
+
+  const [cachedState, updateCachedState] =
+    useSavedCalculatorState(valueWithContext);
+
+  // Calculator input codes (Squiggle strings, selected options in <select>, etc.) are tracked by react-hook-form.
+  // Everything else is tracked in local state and backed up to ViewerContext.
+  const form = useForm<FormShape>({
+    defaultValues: cachedState?.formValues ?? getFormValues(calculator),
+  });
+
+  const [inputValues, setInputValues] = useState<InputValues>(
+    () => cachedState?.inputValues ?? {}
+  );
 
-import { Action, useViewerContext } from "../SquiggleViewer/ViewerProvider.js";
-import {
-  CalculatorAction,
-  CalculatorState,
-  calculatorReducer,
-  hasSameCalculator,
-  initialCalculatorState,
-} from "./calculatorReducer.js";
-import { CalculatorUI } from "./calculatorUI.js";
+  const hashString = useMemo(() => calculator.hashString, [calculator]);
+  const [prevHashString, setPrevHashString] = useState<string | undefined>(
+    cachedState?.hashString
+  );
+
+  // This effect does two things:
+  // - when the calculator is updated, it resets the form
+  // - also, on initial render and on calculator updates, it calculates all input values (which will trigger calculator result calculation)
+  // In both cases, it updates `prevHashString`, which tracks the calculator identity.
+  useEffect(() => {
```

Comment:
This effect is slightly messy, probably can be improved or split in two. I also still don't like the idea of `prevHashString`, but calling `form.reset` on `useEffect(..., [hashString])` would mean that we do the reset on the initial mount, which would _almost_ be ok, but has problems with cachedState.